### PR TITLE
report error on extraneous JSX prop instead of element name

### LINKED ISSUE #33476: Better JSX extraneous property error range
The error range on extraneous JSX properties highlights the component name instead of the offending property:

**Current Behavior**:

```typescript
import * as React from 'react';

function Example(props: {x: number, y: string}) {
    return <>Hello, world.</>;
}

<Example x={3} y="a" z={6} />
 ~~~~~~~
```

**Desired Behavior**:

```typescript
import * as React from 'react';

function Example(props: {x: number, y: string}) {
    return <>Hello, world.</>;
}

<Example x={3} y="a" z={6} />
                     ~~~~~
```

in both cases, the error message is rightly

```
 Type '{ x: number; y: string; z: number; }' is not assignable to type 'IntrinsicAttributes & { x: number; y: string; }'.
   Property 'z' does not exist on type 'IntrinsicAttributes & { x: number; y: string; }'.
```

The situation seems to have been improved from #23117 but I feel that it could still be better by highlighting the offending property in particular.

**TypeScript Version:**  3.6.3 (also verified with 3.7.0-dev.20190914)

**Search Terms:** jsx error range extra property type error message

**Playground Link:** 

[playground](http://www.typescriptlang.org/play/?jsx=2#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdBArgHb7AQtwCiAHsuAA2WABRhiYVAC44Ab14yWTEACMsUADRwAnjNQwowFgHMAvgEo5dODbiUYTKFwA8APgASWQYIhaA7tCCACYAdM4A9K70pgzOfAJgwnC8ALyyAMymOikARMg5cABeaQBsWZFAA)

**Related Issues:** 
#23117

### PR DESCRIPTION
**Pull request by Nathan-Fenner**
Fixes #33476

Before:
```typescript
import * as React from 'react';

function Example(props: {x: number, y: string}) {
    return <>Hello, world.</>;
}

<Example x={3} y="a" z={6} />
 ~~~~~~~
```

After:
```typescript
import * as React from 'react';

function Example(props: {x: number, y: string}) {
    return <>Hello, world.</>;
}

<Example x={3} y="a" z={6} />
                     ~
```

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13041,6 +13041,16 @@ namespace ts {
                                     // JsxAttributes has an object-literal flag and undergo same type-assignablity check as normal object-literal.
                                     // However, using an object-literal error message will be very confusing to the users so we give different a message.
                                     // TODO: Spelling suggestions for excess jsx attributes (needs new diagnostic messages)
+
+                                    if (errorNode && isJsxOpeningLikeElement(errorNode.parent)) {
+                                        const attributes = errorNode.parent.attributes;
+                                        for (const jsxProperty of attributes.properties) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Instead of searching for this, can you just get the `valueDeclaration` of `prop`?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -13041,6 +13041,16 @@ namespace ts {
                                     // JsxAttributes has an object-literal flag and undergo same type-assignablity check as normal object-literal.
                                     // However, using an object-literal error message will be very confusing to the users so we give different a message.
                                     // TODO: Spelling suggestions for excess jsx attributes (needs new diagnostic messages)
+
+                                    if (errorNode && isJsxOpeningLikeElement(errorNode.parent)) {
+                                        const attributes = errorNode.parent.attributes;
+                                        for (const jsxProperty of attributes.properties) {
+                                            if (jsxProperty.kind === SyntaxKind.JsxAttribute && jsxProperty.name.escapedText === prop.escapedName) {
+                                                // Move the error node to the actual JSX property, instead of pointing to the identifier in the JSX element.
+                                                errorNode = jsxProperty;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Instead of the entire attribute, can you error only on the attribute name?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -13041,6 +13041,16 @@ namespace ts {
                                     // JsxAttributes has an object-literal flag and undergo same type-assignablity check as normal object-literal.
                                     // However, using an object-literal error message will be very confusing to the users so we give different a message.
                                     // TODO: Spelling suggestions for excess jsx attributes (needs new diagnostic messages)
+
+                                    if (errorNode && isJsxOpeningLikeElement(errorNode.parent)) {
+                                        const attributes = errorNode.parent.attributes;
+                                        for (const jsxProperty of attributes.properties) {
+                                            if (jsxProperty.kind === SyntaxKind.JsxAttribute && jsxProperty.name.escapedText === prop.escapedName) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: If you still end up having code like this, switch from `jsxProperty` to `attribute` or something similar.

File: src/compiler/checker.ts (hunk 1 above)
Comment (Nathan-Fenner) [reply src/compiler/checker.ts]: I figured there would be a better way of doing this! The AST node API is just so large it was hard to find where to look for it.

File: src/compiler/checker.ts (hunk 2 above)
Comment (Nathan-Fenner) [reply src/compiler/checker.ts]: Yeah, that makes more sense, since it avoids covering up any type errors inside the attribute expression.

**Comment by Nathan-Fenner** [comment]
I switched it to look at `prop.valueDeclaration.name` instead. There's just an odd case that I hadn't considered (but the original approach happened to automatically handle) for extraneous `children`:

```typescript
<ComponentName>extra</ComponentName>
 ~~~~~~~~~~~~~
```

the `name` for the `children` attribute has no location (since the name `"children"` doesn't actually exist in the source). So in this case, the existing behavior is kept, and the underlines stay under `ComponentName`.

File: src/compiler/checker.ts (hunk 3 above)
Comment (Nathan-Fenner) [reply src/compiler/checker.ts]: Removed this so the name is no longer relevant

**Comment by DanielRosenwasser** [comment]
Then I'd just say if you have a `valueDeclaration` that's a JSX Attribute, change the span to the identifier. That's probably the easiest thing.

**Comment by Nathan-Fenner** [comment]
@DanielRosenwasser I updated the PR to check with `isJSXAttribute` now, and highlight attribute names instead of the entire attribute.

I don't understand why the CI checks are failing:

```
The baseline file tsxUnionElementType4.errors.txt has changed.
Error: The baseline file tsxUnionElementType4.errors.txt has changed.
    at writeComparison (src/harness/harness.ts:1722:23)
    at Object.runBaseline (src/harness/harness.ts:1732:13)
    at Object.doErrorBaseline (src/harness/harness.ts:1130:22)
    at CompilerTest.verifyDiagnostics (src/testRunner/compilerRunner.ts:229:26)
    at Context.<anonymous> (src/testRunner/compilerRunner.ts:87:67)
```

is this because there is an approval process for changes to the baseline files? Or is it just not recognizing the checked-in updates to the baselines?

**Comment by DanielRosenwasser** [comment]
Sounds like you need to update the baselines? But I'm not sure why it's failing.

**Comment by Nathan-Fenner** [comment]
Re-ran the tests and was able to update the baselines

**Comment by orta** [comment]
Great work @Nathan-Fenner 👍

**Comment by DanielRosenwasser** [comment]
Thank you!

**Pull request merged**